istioctl/cmd/kubeuninject.go

```diff
@@ -0,0 +1,336 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"github.com/spf13/cobra"
+	"go.uber.org/multierr"
+	"k8s.io/api/batch/v2alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	yamlDecoder "k8s.io/apimachinery/pkg/util/yaml"
+
+	"istio.io/istio/pilot/pkg/kube/inject"
+	"istio.io/pkg/log"
+)
+
+const (
+	proxyContainerName          = "istio-proxy"
+	initContainerName           = "istio-init"
+	enableCoreDumpContainerName = "enable-core-dump"
+	envoyVolumeName             = "istio-envoy"
+	certVolumeName              = "istio-certs"
+	annotationPolicy            = "sidecar.istio.io/inject"
+	sidecarAnnotationPrefix     = "sidecar.istio.io"
+)
+
+func validateUninjectFlags() error {
+	var err error
+
+	if uninjectInFilename == "" {
+		err = multierr.Append(err, errors.New("filename not specified (see --filename or -f)"))
+	}
+	return err
+}
+
+// extractResourceFile uninjects the istio proxy from the specified
+// kubernetes YAML file.
+func extractResourceFile(in io.Reader, out io.Writer) error {
+	reader := yamlDecoder.NewYAMLReader(bufio.NewReaderSize(in, 4096))
+	for {
+		raw, err := reader.Read()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		obj, err := inject.FromRawToObject(raw)
+		if err != nil && !runtime.IsNotRegisteredError(err) {
+			return err
```

Comment:
Add information explaining that the input can't be parsed.  For example `return multierr.Append(fmt.Errorf("cannot parse YAML input"), err)`

Otherwise if someone supplies a non-YAML file they will get a confusing message about JSON.  Users need messages with guidance.

```
echo "hello" | istioctl experimental kube-uninject -f -
Error: error unmarshaling JSON: json: cannot unmarshal string into Go value of type v1.TypeMeta
```